File: src/com/aopphp/go/pointcut/AndPointFilter.java
package com.aopphp.go.pointcut;

import com.jetbrains.php.lang.psi.elements.PhpNamedElement;

import java.util.HashSet;
import java.util.Set;

/**
 * Logic "AND" point filter implementation
 */
public class AndPointFilter implements PointFilter {

    /**
     * Filter kind
     */
    private final Set<KindFilter> kind;

    /**
     * First part of filter
     */
    private PointFilter first;

    /**
     * Second part of filter
     */
    private PointFilter second;

    public AndPointFilter(PointFilter first, PointFilter second) {

        this.first = first;
        this.second = second;

        // intersect kinds
        Set<KindFilter> intersection = new HashSet<KindFilter>(first.getKind());
        intersection.retainAll(second.getKind());

        this.kind = intersection;
    }

    @Override
    public Set<KindFilter> getKind() {
        return kind;
    }

    @Override
    public boolean matches(PhpNamedElement element) {
        return first.matches(element) && second.matches(element);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof AndPointFilter)) return false;

        AndPointFilter that = (AndPointFilter) o;

        if (!kind.equals(that.kind)) return false;
        if (!first.equals(that.first)) return false;
        return second.equals(that.second);
    }

    @Override
    public int hashCode() {
        int result = kind.hashCode();
        result = 31 * result + first.hashCode();
        result = 31 * result + second.hashCode();
        return result;
    }
}

File: src/main/kotlin/com/aopphp/go/pointcut/AndPointFilter.kt
package com.aopphp.go.pointcut

import com.jetbrains.php.lang.psi.elements.PhpNamedElement

class AndPointFilter(private val first: PointFilter, private val second: PointFilter) : PointFilter {
    private val _kind = first.getKind().intersect(second.getKind())

    override fun getKind() = _kind
    override fun matches(element: PhpNamedElement) = first.matches(element) && second.matches(element)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is AndPointFilter) return false
        return _kind == other._kind && first == other.first && second == other.second
    }

    override fun hashCode() = 31 * (31 * _kind.hashCode() + first.hashCode()) + second.hashCode()
}
